sklearn/metrics/pairwise.py

```diff
@@ -671,7 +687,256 @@ def cosine_distances(X, Y=None):
+
+    This implementation modifies the Gower's original similarity measure in
+    the folowing aspects:
+
+    * The values in the original similarity S range between 0 and 1. To
+    guarantee this, it is assumed the numerical features of X and Y are
+    scaled between 0 and 1.
+
+    * Different from the original similarity S, this implementation
+    returns 1-S.
+    """
+    if issparse(X) or issparse(Y):
+        raise TypeError("Gower distance does not support sparse matrices")
+
+    if not isinstance(scale, (bool, list, np.ndarray)):
+        raise TypeError("Parameter scale must be boolean, list, or ndarray")
+
+    if X is None or len(X) == 0:
+        raise ValueError("X can not be None or empty")
+
+    y_is_not_none = Y is not None
+
+    # It is necessary to convert to ndarray in advance to define the dtype
+    # as np.object, otherwise numeric columns will be converted to string
+    # if there are other string columns.
+    if not isinstance(X, np.ndarray):
+        X = np.asarray(X, dtype=np.object)
+
+    if y_is_not_none and not isinstance(Y, np.ndarray):
+        Y = np.asarray(Y, dtype=np.object)
+
+    X, Y = check_pairwise_arrays(X, Y, precomputed=False, dtype=X.dtype,
+                                 force_all_finite=False)
+
+    n_rows, n_cols = X.shape
+
+    categorical_features_obj, categorical_features_num, numerical_features = \
+        _detect_categorical_features(X, categorical_features)
+
+    # Categorical columns
+    X_cat_obj = X[:, categorical_features_obj].astype(np.object)
+    X_cat_num = X[:, categorical_features_num].astype(np.float)
+
+    # Numerical columns
+    X_num = X[:, numerical_features].astype(np.float)
+
+    Y_cat_obj = None
+    Y_cat_num = None
+    Y_num = None
+    if y_is_not_none:
+        Y_cat_obj = Y[:, categorical_features_obj].astype(np.object)
+        Y_cat_num = Y[:, categorical_features_num].astype(np.float)
+        Y_num = Y[:, numerical_features].astype(np.float)
+    else:
+        Y_cat_obj = X_cat_obj
+        Y_cat_num = X_cat_num
+        Y_num = X_num
+
+    num_cols = X_num.shape[1]
+    n_col_num_present = num_cols > 0
+    n_col_cat_obj_present = X_cat_obj.shape[1] > 0
+    n_col_cat_num_present = X_cat_num.shape[1] > 0
+
+    # Calculates the min and max values, and if requested, scale the
+    # input values in order to obtain the distances between 0 and 1,
+    # as proposed by the Gower's paper.
+    if n_col_num_present:
+        process_scale = False
+        if isinstance(scale, bool):
+            process_scale = scale
+        else:
+            if (isinstance(scale, list) and len(scale) != X_num.shape[1]) or\
+               (isinstance(scale, np.ndarray) and
+                    len(scale.flat) != X_num.shape[1]):
+                raise ValueError("Length of scale parameter must be equal "
+                                 "to the number of numerical columns.")
+            process_scale = True
+
+        min = np.nanmin(X_num, axis=0)
```

Comment:
>> If you believe that min is necessary for computation of gower distance (I'm still not entirely convinced), then you need to calculate it in _precompute_metric_params.

We need min together with ranges (contained in scale parameter), in order to define the max, and then use scaler = MinMaxScaler().fit([min, max]).

We agreed before in this PR that the numerical data will be scaled via MinMaxScaler, because it is able to deal properly with nans, and it is an elegant way of code reuse. 

Do you think it is possible to use only the ranges of numerical data with MinMaxScaler?

Reply:
The thing is that here we don't care about the *value* after scaling, we care about the *difference* between two values with the same scaling. The difference between corresponding variables needs to be scaled (before or after subtraction) but the offset makes no difference:

That is, if for some feature we determine a scale_ of `β` and a min_ of `α` as MinMaxScaler defines them, we could either adjust the values from samples 1 and 2, and then subtract them to get the manhattan distance:

> |(βx₁ + α) - (βx₂ + α)|

or we can equivalently subtract and scale them:

> β|x₁ - x₂|

It's the same result, but if we ignore α we can't ensure that `0 ≤  βx₁ + α ≤  1`.

Reply:
In this case the scale parameter won't contain the ranges anymore, but the decimal values to multiply just after the absolute value is retrieved for manhattan calculation, then we don't need MinMaxScaler anymore?

Reply:
No we don't need the scaler implementation. It's a very simple transformation